compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -299,35 +311,60 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
         goal: Goal<'tcx, G>,
     ) -> Vec<Candidate<'tcx>> {
         debug_assert_eq!(goal, self.resolve_vars_if_possible(goal));
+        if let Some(ambig) = self.self_ty_infer_ambiguity_hack(goal) {
+            return ambig;
+        }
+
+        let mut candidates = self.assemble_candidates_via_self_ty(goal);
+
+        self.assemble_blanket_impl_candidates(goal, &mut candidates);
+
+        self.assemble_param_env_candidates(goal, &mut candidates);
 
-        // HACK: `_: Trait` is ambiguous, because it may be satisfied via a builtin rule,
-        // object bound, alias bound, etc. We are unable to determine this until we can at
-        // least structurally resolve the type one layer.
-        if goal.predicate.self_ty().is_ty_var() {
-            return vec![Candidate {
+        candidates
+    }
+
+    fn self_ty_infer_ambiguity_hack<G: GoalKind<'tcx>>(
```

Comment:
nit:

`assemble_self_ty_infer_ambiguity_response`

tbh, at this point, i wouldnt even call it a hack, it seems pretty fundamental to the architecture of the solver?